front/lib/workspace_usage.ts

```diff
@@ -104,6 +150,301 @@ export async function unsafeGetUsageData(
+      [
+        Sequelize.fn(
+          "COUNT",
+          Sequelize.literal('DISTINCT "agent_configuration"."sId"')
+        ),
+        "distinctAgentsEditionsCount",
+      ],
+    ],
+    where: {
+      workspaceId: wId,
+      createdAt: {
+        [Op.gt]: startDate,
+        [Op.lt]: endDate,
+      },
+    },
+    include: [
+      {
+        model: User,
+        as: "user",
+        attributes: [],
+        required: true,
+      },
+    ],
+    raw: true,
+    group: ["authorId", "user.email", "user.firstName", "user.lastName"],
+  });
+  const buildersUsage: builderUsage[] = agentConfigurations.map((result) => {
+    return {
+      userFirstName: (result as unknown as { firstName: string }).firstName,
+      userLastName: (result as unknown as { lastName: string }).lastName,
+      userEmail: (result as unknown as { email: string }).email,
+      agentsEditionsCount: (
+        result as unknown as { agentsEditionsCount: number }
+      ).agentsEditionsCount,
+      distinctAgentsEditionsCount: (
+        result as unknown as { distinctAgentsEditionsCount: number }
+      ).distinctAgentsEditionsCount,
+    };
+  });
+  if (!buildersUsage.length) {
+    return "No data available for the selected period.";
+  }
+  const csvHeader = Object.keys(buildersUsage[0]).join(",") + "\n";
+  const csvContent = buildersUsage
+    .map((row) => Object.values(row).join(","))
+    .join("\n");
+  return csvHeader + csvContent;
+}
+
+export async function getAgentUsage(
+  workspaceId: string,
+  startDate: Date,
+  endDate: Date
+): Promise<string> {
+  const workspace = await Workspace.findOne({
+    where: { sId: workspaceId },
+  });
+  if (!workspace) {
+    throw new Error(`Workspace not found for sId: ${workspaceId}`);
+  }
+  const wId = workspace.id;
+  const mentions = await frontSequelize.query<AgentUsageQueryResult>(
+    `
+    SELECT
+      ac."name",
+      ac."description",
+      CASE
+        WHEN ac."scope" = 'published' THEN 'shared'
+        WHEN ac."scope" = 'private' THEN 'private'
+        ELSE 'company'
+      END AS "settings",
+      ARRAY_AGG(DISTINCT aut."email") AS "authorEmails",
+      COUNT(a."id") AS "messages",
+      COUNT(DISTINCT u."id") AS "distinctUsersReached",
+      MAX(CAST(ac."createdAt" AS DATE)) AS "lastConfiguration"
+    FROM
+      "mentions" a
+      JOIN "messages" m ON a."id" = m."agentMessageId"
+      JOIN "messages" parent ON m."parentId" = parent."id"
+      JOIN "user_messages" um ON um."id" = parent."userMessageId"
+      JOIN "users" u ON um."userId" = u."id"
+      JOIN "agent_configurations" ac ON a."agentConfigurationId" = ac."sId"
+      JOIN "users" aut ON ac."authorId" = aut."id"
+    WHERE
+      a."createdAt" BETWEEN :startDate AND :endDate
+      AND ac."workspaceId" = ${wId}
+      AND ac."status" = 'active'
+      AND ac."scope" != 'private'
+    GROUP BY
+      ac."name",
+      ac."description",
+      ac."scope"
+    ORDER BY
+      "Messages Last 30 Days" DESC;
```

Comment:
```suggestion
      "Messages Last 30 Days" DESC;
```
Does that work on sequelize? or is it a metabase thing?

Reply:
if metabase, that won't work here; if SQL well... TIL

Reply:
It indeed seems to work properly with Sequelize 😇